src/csharp/Grpc.Core.Api/AsyncCallState.cs

```diff
@@ -0,0 +1,92 @@
+#region Copyright notice and license
+
+// Copyright 2015 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+
+using System;
+using System.Threading.Tasks;
+
+namespace Grpc.Core
+{
+    /// <summary>
+    /// Provides an abstraction over the callback providers
+    /// used by AsyncUnaryCall, AsyncDuplexStreamingCall, etc
+    /// </summary>
+    internal /* readonly */ struct AsyncCallState // can be made readonly in C# 7.2
+    {
+        readonly object responseHeadersAsync; // Task<Metadata> or Func<object, Task<Metadata>>
```

Comment:
Possible alternative:
Because after this PR Grpc.Core only uses the newly added Async*Call constructors (and we can easily make grpc-dotnet do the same), it would be fine to make the old constructor allocate more (because it's there only for backward compatibility) - it could actually allocate a new state object which can hold the statusFunc, trailersFunc etc. and thus you could avoid the `object`types members here?